Frameworks/MQTTnet.NetStandard/Server/MqttTopicFilterComparer.cs

```diff
@@ -6,39 +6,115 @@ public static class MqttTopicFilterComparer
     {
         private static readonly char[] TopicLevelSeparator = { '/' };
 
-        public static bool IsMatch(string topic, string filter)
+        public static bool IsMatch(string topic, string sub)
         {
-            if (topic == null) throw new ArgumentNullException(nameof(topic));
-            if (filter == null) throw new ArgumentNullException(nameof(filter));
+            if (string.IsNullOrEmpty(topic)) throw new ArgumentNullException(nameof(topic));
+            if (string.IsNullOrEmpty(sub)) throw new ArgumentNullException(nameof(sub));
 
-            if (string.Equals(topic, filter, StringComparison.Ordinal))
-            {
-                return true;
-            }
 
-            var fragmentsTopic = topic.Split(TopicLevelSeparator, StringSplitOptions.None);
-            var fragmentsFilter = filter.Split(TopicLevelSeparator, StringSplitOptions.None);
+            var spos = 0;
+            var slen = sub.Length;
+            var tpos = 0;
+            var tlen = topic.Length;
+                        
+            var multilevel_wildcard = false;
 
-            for (var i = 0; i < fragmentsFilter.Length; i++)
+            while (spos < slen && tpos <= tlen)
             {
-                switch (fragmentsFilter[i])
+                if (sub[spos] == topic[tpos])
                 {
-                    case "+": continue;
-                    case "#" when i == fragmentsFilter.Length - 1: return true;
+                    if (tpos == tlen - 1)
+                    {
+                        /* Check for e.g. foo matching foo/# */
+                        if (spos == slen - 3
+                                && sub[spos + 1] == '/'
+                                && sub[spos + 2] == '#')
+                        {
+                            multilevel_wildcard = true;
+                            return true;
+                        }
+                    }
+                    spos++;
+                    tpos++;
+                    if (spos == slen && tpos == tlen)
+                    {
+                        return true;
+                    }
+                    else if (tpos == tlen && spos == slen - 1 && sub[spos] == '+')
+                    {
+                        if (spos > 0 && sub[spos - 1] != '/')
+                        {
+                            throw new ArgumentException("Invalid filter string", "filter");
```

Comment:
I don't think that throwing an exception is not a good idea here. If the filter is invalid the server will produce lots of exceptions. In this case I would return "no match" and write a warning to the log. Or the better approach is valdiating the topic while being subscribed and return a failure to the client. Then this code needs no exceptions.